File: retaguarda-software-house/java/retaguarda_sh/src/main/java/com/t2ti/retaguarda_sh/service/acbr/AcbrMonitorService.java
/*******************************************************************************
Title: T2Ti ERP 3.0                                                                
Description: Service relacionado à tabela [NFE_CONFIGURACAO] 
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2021 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                     
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,                    
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************/
package com.t2ti.retaguarda_sh.service.acbr;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Base64;
import java.util.List;
import java.util.Scanner;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;

import org.apache.commons.io.IOUtils;
import org.ini4j.Wini;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.zeroturnaround.zip.ZipUtil;

import com.t2ti.retaguarda_sh.model.cadastros.Empresa;
import com.t2ti.retaguarda_sh.model.transiente.ObjetoNfe;
import com.t2ti.retaguarda_sh.service.cadastros.EmpresaService;
import com.t2ti.retaguarda_sh.util.Biblioteca;

@Service
public class AcbrMonitorService {

	@Autowired
	private EmpresaService empresaService;
	
	@PersistenceContext
    private EntityManager entityManager;	

	private String caminhoComCnpj = "";	
	
	@SuppressWarnings("resource")
	public String emitirNfce(String numero, String cnpj, String nfceIni) throws IOException, InterruptedException {
		// configurações
		caminhoComCnpj = "C:\\ACBrMonitor\\" + cnpj + "\\";
		
		// salva o arquivo INI em disco
		String caminhoArquivoIniNfce = caminhoComCnpj + "ini\\nfce\\" + numero + ".ini";
		byte[] nfceIniBytes = Base64.getDecoder().decode(nfceIni);
		FileOutputStream fileOutputStream = new FileOutputStream(caminhoArquivoIniNfce);
		fileOutputStream.write(nfceIniBytes);
		fileOutputStream.close();

		// chama o método para criar a nota
		criarNFe(caminhoArquivoIniNfce);
		// pega o caminho do XML criado
		String caminhoArquivoXml = pegarRetornoSaida("ARQUIVO-XML");
		// chama o método para criar e enviar a nota
		enviarNFe(caminhoArquivoXml);
		String retorno = pegarRetornoSaida("Envio");
		if (!retorno.contains("ERRO")) {
			// chama o método para gerar o PDF
		    imprimirDanfe(caminhoArquivoXml);
		    // captura o retorno do arquivo SAI
		    retorno = pegarRetornoSaida("ARQUIVO-PDF");
		}		
		return retorno;
	}
	
	@SuppressWarnings("resource")
	public String emitirNfceContingencia(String numero, String cnpj, String nfceIni) throws IOException, InterruptedException {
		// configurações
		caminhoComCnpj = "C:\\ACBrMonitor\\" + cnpj + "\\";
		// salva o arquivo INI em disco
		String caminhoArquivoIniNfce = caminhoComCnpj + "ini\\nfce\\" + numero + ".ini";
		byte[] nfceIniBytes = Base64.getDecoder().decode(nfceIni);

		FileOutputStream fileOutputStream = new FileOutputStream(caminhoArquivoIniNfce);
		fileOutputStream.write(nfceIniBytes);
		fileOutputStream.close();
		
		// passa para o modo de emissão off-line
		passarParaModoOffLine();
		// chama o método para criar a nota
		criarNFe(caminhoArquivoIniNfce);
		// pega o caminho do XML criado
		String caminhoArquivoXml = pegarRetornoSaida("ARQUIVO-XML");
		// chama o método para gerar o PDF
		imprimirDanfe(caminhoArquivoXml);
		// captura o retorno do arquivo SAI
		String retorno = pegarRetornoSaida("ARQUIVO-PDF");
		// passa para o modo de emissão on-line
		passarParaModoOnLine();
	  
		return retorno;
	}
	
	public String transmitirNfceContingenciada(String chave, String cnpj) throws IOException, InterruptedException {
		// configurações
		caminhoComCnpj = "C:\\ACBrMonitor\\" + cnpj + "\\";
		String caminhoArquivoXml = caminhoComCnpj + "LOG_NFe\\" + chave + "-nfe.xml";
		// chama o método para criar e enviar a nota
		enviarNFe(caminhoArquivoXml);		
		String retorno = pegarRetornoSaida("Envio");
		if (!retorno.contains("ERRO")) {
			// chama o método para gerar o PDF
		    imprimirDanfe(caminhoArquivoXml);
		    // captura o retorno do arquivo SAI
		    retorno = pegarRetornoSaida("ARQUIVO-PDF");
		}		
		return retorno;
	}

	public String tratarNotaAnteriorContingencia(ObjetoNfe objetoNfe) throws IOException, InterruptedException {
		// configurações
		caminhoComCnpj = "C:\\ACBrMonitor\\" + objetoNfe.getCnpj() + "\\";
		String caminhoArquivoXml = caminhoComCnpj + "LOG_NFe\\" + objetoNfe.getChaveAcesso() + "-nfe.xml";

		// vamos verificar o status da nota
		apagarArquivoSaida();
		gerarArquivoEntrada("NFE.ConsultarNFe(" + caminhoArquivoXml + ")");
		aguardarArquivoSaida();
		
		String retorno = pegarRetornoSaida("Consulta");
		if (!retorno.contains("ERRO")) {
			// se a nota anterior foi emitida = cancela. senão = inutiliza.
			if (retorno.contains("Autorizado")) {
				retorno = cancelarNfce(objetoNfe);
			} else {
				retorno = inutilizarNumero(objetoNfe);
			}
		}
		
		return retorno;
	}
	
	public String inutilizarNumero(ObjetoNfe objetoNfe) throws IOException, InterruptedException {
		// configurações
		caminhoComCnpj = "C:\\ACBrMonitor\\" + objetoNfe.getCnpj() + "\\";
		apagarArquivoSaida();
		
		gerarArquivoEntrada("NFE.InutilizarNFe("
	      +objetoNfe.getCnpj() +", "
	      +objetoNfe.getJustificativa() +", "
	      +objetoNfe.getAno() +", "
	      +objetoNfe.getModelo() +", "
	      +objetoNfe.getSerie() +", "
	      +objetoNfe.getNumeroInicial() +", "
	      +objetoNfe.getNumeroFinal() +")");

 	    aguardarArquivoSaida();
		return pegarRetornoSaida("Inutilizacao");
	}

	public String cancelarNfce(ObjetoNfe objetoNfe) throws IOException, InterruptedException {
		caminhoComCnpj = "C:\\ACBrMonitor\\" + objetoNfe.getCnpj() + "\\";
		apagarArquivoSaida();
		gerarArquivoEntrada("NFe.CANCELARNFE(" + objetoNfe.getChaveAcesso() + ", " + objetoNfe.getJustificativa() + ", " + objetoNfe.getCnpj() + ")");

 	    aguardarArquivoSaida();
		return pegarRetornoSaida("Cancelamento");
	}

	public String gerarPdfDanfeNfce(String chave, String cnpj) throws IOException, InterruptedException {
		// configurações
		caminhoComCnpj = "C:\\ACBrMonitor\\" + cnpj + "\\";
		// pega o caminho do arquivo XML da nota em contingência
		String caminhoArquivoXml = caminhoComCnpj + "LOG_NFe\\" + chave + "-nfe.xml";
		// chama o método para gerar o PDF
		imprimirDanfe(caminhoArquivoXml);
		// captura o retorno do arquivo SAI
		return pegarRetornoSaida("ARQUIVO-PDF");	
	}
	
	public void criarNFe(String caminhoArquivoIniNfce) throws IOException, InterruptedException {
		apagarArquivoSaida();
		gerarArquivoEntrada("NFE.CriarNFe(" + caminhoArquivoIniNfce + ")"); 
		aguardarArquivoSaida();
	}

	public void enviarNFe(String caminhoArquivoXml) throws IOException, InterruptedException {
		apagarArquivoSaida();
		gerarArquivoEntrada("NFE.EnviarNFe(" + caminhoArquivoXml + ", 001, , , , 1, , )"); 
		aguardarArquivoSaida();
	}
	
	public void imprimirDanfe(String caminhoArquivoXml) throws IOException, InterruptedException {
		apagarArquivoSaida();
		gerarArquivoEntrada("NFE.ImprimirDANFEPDF(" + caminhoArquivoXml + ", , , 1,)"); 
		aguardarArquivoSaida();
	}
	
	public boolean passarParaModoOffLine() throws IOException, InterruptedException {
		apagarArquivoSaida();
		gerarArquivoEntrada("NFE.SetFormaEmissao(9)"); // 9=offline
		return aguardarArquivoSaida();
	}

	public boolean passarParaModoOnLine() throws IOException, InterruptedException {
		apagarArquivoSaida();
		gerarArquivoEntrada("NFE.SetFormaEmissao(1)"); // 1=normal
		return aguardarArquivoSaida();
	}
	
	public boolean gerarZipArquivosXml(String periodo, String cnpj) {
		String filtro = "CNPJ = '" + cnpj + "'";
		Empresa empresa = empresaService.consultarObjetoFiltro(filtro);
		if (empresa != null) {
			ZipUtil.pack(new File("C:\\ACBrMonitor\\" + cnpj + "\\DFes\\" + periodo), new File("C:\\ACBrMonitor\\" + cnpj + "\\NotasFiscaisNFCe_" + periodo + ".zip"));
			return true; 
		} else {
			return false;
		}
	}
	
	@SuppressWarnings("resource")
	public void atualizarCertificado(String certificadoBase64, String senha, String cnpj) throws FileNotFoundException, IOException, InterruptedException {
		String filtro = "CNPJ = '" + cnpj + "'";
		Empresa empresa = empresaService.consultarObjetoFiltro(filtro);
		if (empresa != null) {
		    // configura os caminhos
			String caminhoComCnpj = "C:\\ACBrMonitor\\" + empresa.getCnpj() + '\\';
		    String caminhoArquivoCertificado = caminhoComCnpj + empresa.getCnpj() + ".pfx";

		    // converte e salva o arquivo do certificado em disco
			byte[] certificadoBytes = Base64.getDecoder().decode(certificadoBase64);
			FileOutputStream fileOutputStream = new FileOutputStream(caminhoArquivoCertificado);
			fileOutputStream.write(certificadoBytes);
			fileOutputStream.close();
			
		    apagarArquivoSaida();
		    gerarArquivoEntrada("NFE.SetCertificado(" + caminhoArquivoCertificado + "," + senha + ")");
		    aguardarArquivoSaida();
			
		    Biblioteca.killTask("ACBrMonitor_" + empresa.getCnpj() + ".exe");
		    String caminhoExecutavel = caminhoComCnpj + "ACBrMonitor_" + empresa.getCnpj() + ".exe";
		    Runtime.getRuntime().exec(caminhoExecutavel);
		}		
	}
	
	private void gerarArquivoEntrada(String comando) throws IOException
	{
		FileWriter ArquivoEntrada = new FileWriter(caminhoComCnpj + "ent.txt");
		PrintWriter gravarArquivo = new PrintWriter(ArquivoEntrada);
		gravarArquivo.printf(comando);
		ArquivoEntrada.close();
	}
	
	private String pegarRetornoSaida(String operacao) throws IOException {
		String retorno = "";
		Wini arquivoSaida = new Wini(new File(caminhoComCnpj + "sai.txt"));

		// carrega o conteúdo completo do arquivo
		FileInputStream arquivoCompleto = new FileInputStream(caminhoComCnpj + "sai.txt"); 
		Scanner scanner = new Scanner(arquivoCompleto);		

		String codigoStatus = arquivoSaida.get(operacao, "CStat");
		String motivo = arquivoSaida.get(operacao, "XMotivo");
		
		String caminhoArquivoXml = "";
		
		try {			
			if (operacao.equals("ARQUIVO-XML")) {
	            caminhoArquivoXml = scanner.nextLine();
	            caminhoArquivoXml = caminhoArquivoXml.replaceAll("OK: ", "");
	            caminhoArquivoXml = caminhoArquivoXml.trim();
	            return caminhoArquivoXml; 
			} else if (operacao.equals("ARQUIVO-PDF")) {
				retorno = IOUtils.toString(arquivoCompleto, StandardCharsets.UTF_8.name());
				return retorno.replaceAll("OK: Arquivo criado em: ", "").trim();					
			} else if (operacao.equals("Envio")) {
				retorno = motivo;
			} else if (operacao.equals("Cancelamento")) {
				retorno = motivo;
			} else if (operacao.equals("Consulta")) {
				retorno = motivo;
			} else if (operacao.equals("Inutilizacao")) {
				return IOUtils.toString(arquivoCompleto, StandardCharsets.UTF_8.name());
			}
			
			List<String> listaStatus = Arrays.asList("", "100" , "102", "135"); // se o status não for um dos que estiverem nessa lista, vamos retornar um erro.
			
			if (!listaStatus.contains(codigoStatus)) {
				return "[ERRO] - [" + codigoStatus + "] " + motivo;
			}
		} finally {
			scanner.close(); 
		}
		return retorno;
	}
	
	private boolean apagarArquivoSaida() {
		File file = new File(caminhoComCnpj + "sai.txt"); 
		return file.delete();		
	}
	
	private boolean aguardarArquivoSaida() throws InterruptedException {
		int tempoEspera = 0;
		
		File file = new File(caminhoComCnpj + "sai.txt"); 
		while (!file.exists()) {
			Thread.sleep(1000);
			tempoEspera++;
			
			if (tempoEspera > 30) {
				return false;
			}
		}
		
		return true;
	}	
}

File: retaguarda-software-house/node/retaguarda-sh/src/acbr/acbr-monitor/acbr-monitor.service.ts
/*******************************************************************************
Title: T2Ti ERP 3.0                                                                
Description: Service relacionado ao ACBrMonitor
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2021 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                     
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,                    
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************/
import { Injectable } from '@nestjs/common';
import { getConnection } from 'typeorm';
import { Empresa } from '../../entities-export';
import { Biblioteca } from '../../util/biblioteca';
import { ObjetoNfe } from '../../util/objeto.nfe';
import { Ini } from '../../util/ini';
import * as fs from "fs";

@Injectable()
export class AcbrMonitorService {

    caminhoComCnpj: string = "";	

	async emitirNfce(numero: string, cnpj: string, nfceIni: string): Promise<string> {
		// configurações
		this.caminhoComCnpj = "C:\\ACBrMonitor\\" + cnpj + "\\";
		
		// salva o arquivo INI em disco
		const caminhoArquivoIniNfce = this.caminhoComCnpj + "ini\\nfce\\" + numero + ".ini";
        let buff = Buffer.from(nfceIni, 'base64');
        fs.writeFileSync(caminhoArquivoIniNfce, buff);

		// chama o método para criar a nota
		await this.criarNFe(caminhoArquivoIniNfce);
		// pega o caminho do XML criado
		let caminhoArquivoXml = await this.pegarRetornoSaida("ARQUIVO-XML");
		// chama o método para criar e enviar a nota
		await this.enviarNFe(caminhoArquivoXml);
		let retorno = await this.pegarRetornoSaida("Envio");
		if (!retorno.includes("ERRO")) {
			// chama o método para gerar o PDF
		    await this.imprimirDanfe(caminhoArquivoXml);
		    // captura o retorno do arquivo SAI
		    retorno = await this.pegarRetornoSaida("ARQUIVO-PDF");
		}		
		return retorno;
    }

	async emitirNfceContingencia(numero: string, cnpj: string, nfceIni: string): Promise<string> {
		// configurações
		this.caminhoComCnpj = "C:\\ACBrMonitor\\" + cnpj + "\\";

		// salva o arquivo INI em disco
		const caminhoArquivoIniNfce = this.caminhoComCnpj + "ini\\nfce\\" + numero + ".ini";
        let buff = Buffer.from(nfceIni, 'base64');
        fs.writeFileSync(caminhoArquivoIniNfce, buff);
		
		// passa para o modo de emissão off-line
		await this.passarParaModoOffLine();
		// chama o método para criar a nota
		await this.criarNFe(caminhoArquivoIniNfce);
		// pega o caminho do XML criado
		const caminhoArquivoXml = await this.pegarRetornoSaida("ARQUIVO-XML");
		// chama o método para gerar o PDF
		await this.imprimirDanfe(caminhoArquivoXml);
		// captura o retorno do arquivo SAI
		let retorno = await this.pegarRetornoSaida("ARQUIVO-PDF");
		// passa para o modo de emissão on-line
		await this.passarParaModoOnLine();
	  
		return retorno;    
    }

	async transmitirNfceContingenciada(chave: string, cnpj: string): Promise<string> {
		// configurações
		this.caminhoComCnpj = "C:\\ACBrMonitor\\" + cnpj + "\\";
		const caminhoArquivoXml = this.caminhoComCnpj + "LOG_NFe\\" + chave + "-nfe.xml";
		// chama o método para criar e enviar a nota
		await this.enviarNFe(caminhoArquivoXml);		
		let retorno = await this.pegarRetornoSaida("Envio");
		if (!retorno.includes("ERRO")) {
			// chama o método para gerar o PDF
		    await this.imprimirDanfe(caminhoArquivoXml);
		    // captura o retorno do arquivo SAI
		    retorno = await this.pegarRetornoSaida("ARQUIVO-PDF");
		}		
		return retorno;
    }

	async tratarNotaAnteriorContingencia(objetoNfe: ObjetoNfe): Promise<string> {
		// configurações
		this.caminhoComCnpj = "C:\\ACBrMonitor\\" + objetoNfe.cnpj + "\\";
		const caminhoArquivoXml = this.caminhoComCnpj + "LOG_NFe\\" + objetoNfe.chaveAcesso + "-nfe.xml";

		// vamos verificar o status da nota
		await this.apagarArquivoSaida();
		await this.gerarArquivoEntrada("NFE.ConsultarNFe(" + caminhoArquivoXml + ")");
		await this.aguardarArquivoSaida();
		
		let retorno = await this.pegarRetornoSaida("Consulta");
		if (!retorno.includes("ERRO")) {
			// se a nota anterior foi emitida = cancela. senão = inutiliza.
			if (retorno.includes("Autorizado")) {
				retorno = await this.cancelarNfce(objetoNfe);
			} else {
				retorno = await this.inutilizarNumero(objetoNfe);
			}
		}
		
		return retorno;
    }

	async inutilizarNumero(objetoNfe: ObjetoNfe): Promise<string> {
		// configurações
		this.caminhoComCnpj = "C:\\ACBrMonitor\\" + objetoNfe.cnpj + "\\";
		await this.apagarArquivoSaida();
		
		await this.gerarArquivoEntrada("NFE.InutilizarNFe("
            +objetoNfe.cnpj +", "
            +objetoNfe.justificativa +", "
            +objetoNfe.ano +", "
            +objetoNfe.modelo +", "
            +objetoNfe.serie +", "
            +objetoNfe.numeroInicial +", "
            +objetoNfe.numeroFinal +")");

		await this.aguardarArquivoSaida();		
		const retorno = await this.pegarRetornoSaida("Inutilizacao");
		return retorno;
    }

	async cancelarNfce(objetoNfe: ObjetoNfe): Promise<string> {
		this.caminhoComCnpj = "C:\\ACBrMonitor\\" + objetoNfe.cnpj + "\\";
		await this.apagarArquivoSaida();
		await this.gerarArquivoEntrada("NFe.CANCELARNFE(" + objetoNfe.chaveAcesso + ", " + objetoNfe.justificativa + ", " + objetoNfe.cnpj + ")");
		await this.aguardarArquivoSaida();		
		const retorno = await this.pegarRetornoSaida("Cancelamento");
		return retorno;
    }

	async gerarPdfDanfeNfce(chave: string, cnpj: string): Promise<string> {
		// configurações
		this.caminhoComCnpj = "C:\\ACBrMonitor\\" + cnpj + "\\";		
		// pega o caminho do arquivo XML da nota em contingência
		const caminhoArquivoXml = this.caminhoComCnpj + "LOG_NFe\\" + chave + "-nfe.xml";
		// chama o método para gerar o PDF
		await this.imprimirDanfe(caminhoArquivoXml);
		// captura o retorno do arquivo SAI
		return this.pegarRetornoSaida("ARQUIVO-PDF");	
    }

    async criarNFe(caminhoArquivoIniNfce: string) {
		await this.apagarArquivoSaida();
		await this.gerarArquivoEntrada("NFE.CriarNFe(" + caminhoArquivoIniNfce + ")"); 
		await this.aguardarArquivoSaida();
    }

    async enviarNFe(caminhoArquivoXml: string) {
		await this.apagarArquivoSaida();
		await this.gerarArquivoEntrada("NFE.EnviarNFe(" + caminhoArquivoXml + ", 001, , , , 1, , )"); 
		await this.aguardarArquivoSaida();
    }

	async imprimirDanfe(caminhoArquivoXml: string) {
		await this.apagarArquivoSaida();
		await this.gerarArquivoEntrada("NFE.ImprimirDANFEPDF(" + caminhoArquivoXml + ", , , 1,)"); 
		await this.aguardarArquivoSaida();
	}

	async passarParaModoOffLine(): Promise<boolean> {
		await this.apagarArquivoSaida();
		await this.gerarArquivoEntrada("NFE.SetFormaEmissao(9)"); // 9=offline
		return this.aguardarArquivoSaida();
	}

	async passarParaModoOnLine(): Promise<boolean> {
		await this.apagarArquivoSaida();
		await this.gerarArquivoEntrada("NFE.SetFormaEmissao(1)"); // 1=normal
		return this.aguardarArquivoSaida();
	}

    async gerarZipArquivosXml(periodo: string, cnpj: string): Promise<boolean> {
        const connection = getConnection();
        const queryRunner = connection.createQueryRunner();  
        await queryRunner.connect();
        await queryRunner.startTransaction();

        let empresa = await connection.manager.findOne(Empresa, { where: { cnpj: cnpj }} );
        if (empresa != null) {
            const pasta = 'C:\\ACBrMonitor\\' + cnpj + '\\DFes\\' + periodo;
            const arquivoZip = 'C:\\ACBrMonitor\\' + cnpj + '\\NotasFiscaisNFCe_' + periodo + '.zip';
            var zipper = require('zip-local');
            zipper.sync.zip(pasta).compress().save(arquivoZip);
            return true;
        } else {
            return false;
        }
    }

    async atualizarCertificado(certificadoBase64: string, senha: string, cnpj: string) {
        const connection = getConnection();
        const queryRunner = connection.createQueryRunner();  
        await queryRunner.connect();
        await queryRunner.startTransaction();

        const empresa = await connection.manager.findOne(Empresa, { where: { cnpj: cnpj }} );
        if (empresa != null) {
		    // configura os caminhos
            this.caminhoComCnpj = 'C:\\ACBrMonitor\\' + empresa.cnpj + '\\';
		    const caminhoArquivoCertificado = this.caminhoComCnpj + empresa.cnpj + ".pfx";
		    
		    // converte e salva o arquivo do certificado em disco
            let buff = Buffer.from(certificadoBase64, 'base64');
            fs.writeFileSync(caminhoArquivoCertificado, buff);

		    await this.apagarArquivoSaida();
		    await this.gerarArquivoEntrada("NFE.SetCertificado(" + caminhoArquivoCertificado + "," + senha + ")");
		    await this.aguardarArquivoSaida();
						
            Biblioteca.killTask('ACBrMonitor_' + empresa.cnpj + '.exe');
            let caminhoExecutavel = this.caminhoComCnpj + 'ACBrMonitor_' + empresa.cnpj + '.exe';
            require('child_process').exec(caminhoExecutavel);        
        }
    }    

	async gerarArquivoEntrada(comando: string) {
		const nomeArquivo = this.caminhoComCnpj + "ent.txt";
		fs.writeFileSync(nomeArquivo, comando);
	}

	async pegarRetornoSaida(operacao: string): Promise<string> {
		let retorno: string = "";

		const nomeArquivo = this.caminhoComCnpj + "sai.txt";
        const arquivoSaida = new Ini(fs.readFileSync(nomeArquivo, {encoding: 'utf8'}));

        const arquivoCompleto = fs.readFileSync(this.caminhoComCnpj + "sai.txt", {encoding: 'utf8'}).toString();

        let codigoStatus: string = arquivoSaida.getKeyIfExists('CStat') == null ? '' : arquivoSaida.getKeyIfExists('CStat').val;
        let motivo: string = arquivoSaida.getKeyIfExists('XMotivo') == null ? '' : arquivoSaida.getKeyIfExists('XMotivo').val;

		let caminhoArquivoXml = "";

        if (operacao == "ARQUIVO-XML") {
            caminhoArquivoXml = arquivoCompleto.replace("OK: ", "");
            return caminhoArquivoXml.trim();
        } else if (operacao == "ARQUIVO-PDF") {
            let retorno = arquivoCompleto.replace("OK: Arquivo criado em: ", "");
            return retorno.trim();
        } else if (operacao == "Envio") {
            retorno = motivo;
        } else if (operacao == "Cancelamento") {
            retorno = motivo;
        } else if (operacao == "Consulta") {
            retorno = motivo;
        } else if (operacao == "Inutilizacao") {
            return arquivoCompleto;
        }
        
        const listaStatus = ["", "100" , "102", "135"]; // se o status não for um dos que estiverem nessa lista, vamos retornar um erro.
        
        if (listaStatus.filter(function (str) { return str.includes(codigoStatus); }).length == 0) {
            return "[ERRO] - [" + codigoStatus + "] " + motivo;
        }

		return retorno;
	}

	async apagarArquivoSaida() {
		const nomeArquivo = this.caminhoComCnpj + "sai.txt";
        if (fs.existsSync(nomeArquivo)) {
            fs.unlinkSync(nomeArquivo);
        }
	}

	async aguardarArquivoSaida(): Promise<boolean> {
		const nomeArquivo = this.caminhoComCnpj + "sai.txt";
        let tempoEspera = 0;
        while (!fs.existsSync(nomeArquivo)) { 
            await new Promise(resolve => setTimeout(resolve, 1000));
			tempoEspera++;
			
			if (tempoEspera > 30) {
				return false;
			}
        }
        return true;
	}


}